apps/api/src/common/authority-checker.service.ts

```diff
@@ -0,0 +1,380 @@
+import { PrismaClient, Authority, Workspace } from '@prisma/client'
+import { VariableWithProjectAndVersion } from 'src/variable/variable.types'
+import {
+  BadRequestException,
+  Injectable,
+  NotFoundException,
+  UnauthorizedException
+} from '@nestjs/common'
+import getCollectiveProjectAuthorities from './get-collective-project-authorities'
+import getCollectiveWorkspaceAuthorities from './get-collective-workspace-authorities'
+import { EnvironmentWithProject } from 'src/environment/environment.types'
+import { ProjectWithSecrets } from 'src/project/project.types'
+import { SecretWithProjectAndVersion } from 'src/secret/secret.types'
+
+export interface AuthorityInput {
+  userId: string
+  entity: {
+    id?: string
+    name?: string
+  }
+  authority: Authority
+  prisma: PrismaClient
+}
+
+@Injectable()
+export class AuthorityCheckerService {
```

Comment:
I guess since we are not using any dependency injection as such, we can refactor this to a complex object. The object will then contain the functions?

Or maybe we can let the old file structure be? since that was closely aligned to SRP.

Reply:
After implementing the requested logging into the authority checks, this service has to be Injectable.

I really don't think this can or should be done using standalone function exports. The responsibility of authority checks should be handled by a service in a central place (as a beginning point at least).

I didn't dig too deep into the business logic, but there might be room for the generalization of this functionality if there was a generic base repository & separate repositories that extend that generic base repository for the handling of these entities.
I'd say the whole authentication - authorization could be moved out into their separate services, and maybe even handled by a decorator on the controller level after that, which would simplify this by a lot.